noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr

```diff
@@ -0,0 +1,266 @@
+use dep::protocol_types::{
+    address::AztecAddress,
+    constants::GENERATOR_INDEX__NOTE_HASH,
+    hash::poseidon2_hash_with_separator,
+    traits::{Hash, Serialize},
+};
+
+use crate::{
+    context::{PrivateContext, UtilityContext},
+    keys::getters::{get_nsk_app, get_public_keys},
+    oracle::notes::check_nullifier_exists,
+    state_vars::storage::HasStorageSlot,
+};
+
+mod test;
+
+/// # SingleUseClaim
+///
+/// A private state variable type that represents a right a user can exercise at most once.
+///
+/// You can declare a `SingleUseClaim` state variable within your contract's `#[storage]` struct:
+///
+/// `your_action: SingleUseClaim<ClaimIdentifier, Context>`
+///
+/// Where `ClaimIdentifier` is a generic argument, typically a struct provided by the contract developer,
+/// whose only requirement is to implement the `Serializable` trait.
+///
+/// A `ClaimIdentifier` uniquely describes a single-use entitlement in the application domain
+/// (e.g. which voting session, which credential, which proposal, etc).
+///
+/// ## Functions
+/// - `at(owner: AztecAddress)`: scopes the claim to the given owner,
+/// just like `Map#at(address)` gets state specifically related to a given address).
+/// Returns a `SingleUseClaimWithOwner`.
+///
+/// ### Functions scoped to an owner (i.e. `SingleUseClaimWithOwner` functions)
+///
+/// #### Private functions
+/// - `claim(claim_id: ClaimIdentifier)`: records that the owner has exercise the given claim. Emits a nullifier.
+/// - `assert_claimed(claim_id: ClaimIdentifier)`: fails if the owner has not yet exercised the given claim.
+/// Validates existence of a nullifier.
+///
+/// #### Utility functions
+/// - `is_done(claim_id: ClaimIdentifier) -> bool`: returns whether the owner has exercised the given claim.
+/// Checks existence of a nullifier.
```

Comment:
This is not needed because the docsite will automatically pull these by parsing the source code. These explanations would go in the docs for each.